records/2022-07-20.security.prefix-api-token.md

```diff
@@ -0,0 +1,66 @@
+# Prefixing of API tokens
+
+Date: 2022-07-20
+
+## Status
+
+**DRAFT**
+
+## Context
+
+The mishandling of application secrets has lead to many security incidents in companies big and small. Given the inherent risk in how users store these secrets, it is important that we provide mechanisms to protect our users from doing so improperly.  Taking a cue from [Github](https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/), [Slack](https://api.slack.com/authentication/token-types), and [Stripe](https://stripe.com/docs/api/authentication), we see that token prefixes are an effective technique to help us ensure our tokens don't end up in public repositories.
+
+By combining a known prefix with initiatives such as Github's [secret scanning program](https://docs.github.com/en/developers/overview/secret-scanning-partner-program), we can be alerted if one of our users mistakenly stores their API keys in the open.
+
+## Implementation steps
+
+- Add a known prefix to our API keys in order to be able to easily identify them
+- Register with Github's secret scanning program and create an endpoint for them to report issues back to us
+- Give users a grace period to generate new keys
+  - After grace period, enforce a rule that ensures all API keys contain our prefix
+
+### Old API key deprecation process
+
+In order to deprecate old keys each user must be contacted to let them know they need to regenerate their keys and update their applications.  Until all users have regenerated their API keys, we can't start enforcing the existence of the prefix
+
+#### Process outline
+1. Contact clients with non-prefixed keys (based on key creation date); give them a deadline (TBD) and instructions on how to re-generate their API keys
```

Comment:
@yaelberger-commits can I get your feed back on the process for deprecating the older un-prefixed keys?